src/System.Private.CoreLib/src/System/Threading/SynchronizationContext.WinRT.cs

```diff
@@ -0,0 +1,119 @@
+        {
+            get
+            {
+                return s_current ?? GetWinRTContext();
+            }
+        }
+
+        //
+        // It's important that we always return the same SynchronizationContext object for any particular ICoreDispatcher
+        // object, as long as any existing instance is still reachable.  This allows reference equality checks against the 
+        // SynchronizationContext to determine if two instances represent the same dispatcher.  Async frameworks rely on this.
+        // To accomplish this, we use a ConditionalWeakTable to track which instances of WinRTSynchronizationContext are bound
+        // to each ICoreDispatcher instance.
+        //
+        private static readonly ConditionalWeakTable<Object, WinRTSynchronizationContext> s_winRTContextCache =
+            new ConditionalWeakTable<Object, WinRTSynchronizationContext>();
+
+        private static SynchronizationContext GetWinRTContext()
+        {
+            var dispatcher = WinRTInterop.Callbacks.GetCurrentCoreDispatcher();
+            if (dispatcher == null)
+                return null;
+
+            return s_winRTContextCache.GetValue(dispatcher, _dispatcher => new WinRTSynchronizationContext(_dispatcher));
+        }
+    }
+
+
+    internal sealed class WinRTSynchronizationContext : SynchronizationContext
+    {
+        private readonly Object m_dispatcher;
+
+        internal WinRTSynchronizationContext(Object dispatcher)
+        {
+            m_dispatcher = dispatcher;
+        }
+
+        private class Invoker
+        {
+            private readonly SendOrPostCallback m_callback;
+            private readonly object m_state;
+
+            public static readonly Action<object> InvokeDelegate = Invoke;
+
+            public Invoker(SendOrPostCallback callback, object state)
+            {
+                m_callback = callback;
+                m_state = state;
+            }
+
+            public static void Invoke(object thisObj)
+            {
+                var invoker = (Invoker)thisObj;
+                invoker.InvokeCore();
+            }
+
+            private void InvokeCore()
+            {
+                SynchronizationContext prevSyncCtx = SynchronizationContext.CurrentExplicit;
+                try
+                {
+                    m_callback(m_state);
+                }
+                catch (Exception ex)
+                {
+                    //
+                    // If we let exceptions propagate to CoreDispatcher, it will swallow them with the idea that someone will
+                    // observe them later using the IAsyncInfo returned by CoreDispatcher.RunAsync.  However, we ignore
+                    // that IAsyncInfo, because there's nothing Post can do with it (since Post returns void).
+                    // So, we report these as unhandled exceptions.
+                    //
+                    RuntimeAugments.ReportUnhandledException(ex);
```

Comment:
Just kind of related to your review: We'll probably be able to clean up/delete a bunch of the `RuntimeAugments` now that we moved things into CoreLib.